exonum/src/runtime/mod.rs

```diff
@@ -294,10 +333,12 @@ pub trait Runtime: Send + fmt::Debug + 'static {
     ///
     /// # Return value
     ///
-    /// Any error or panic returned from this method should be considered fatal. There are edge
-    /// cases where the returned error does not stop the enclosing process (e.g.,
-    /// if several alternative initial service configurations are tried), but as a rule of thumb,
-    /// a `Runtime` should not return an error or panic here unless it wants the node to stop forever.
+    /// An error or panic returned from this method will not be processed and will lead
+    /// to the node stopping. A runtime should only return an error / panic if the error is local
```

Comment:
Am I right that the Java runtime that performs _some_ user and framework code (= same on all nodes and hopefully mostly deterministic — we can't say that of all the framework and library code) that is not perfomed in `start_adding_service` (but is not considered likely to fail) violates this restriction? But respects

> The runtime should commit long-term resources for the service after a `commit_service()` call.

Is that fine given small/unknown likelyhood? Or shall we attempt to do everything in `start_adding_services` and then just flip a switch (that would complicate things with clean ups in after_commit, and, if we need that, a restriction of API operation of a newly added service before commit_service; oh, and with no `start_adding_` for restarts, it will have to operate differently for newly added and restarted services)?

Reply:
> Am I right that the Java runtime [...] violates this restriction?

There is no 100% non-violable restriction here, hence the use of "should" instead of "must" as per [RFC 2119](https://tools.ietf.org/html/rfc2119). The balance between delaying resource commitment and potentially ending up with the non-functional blockchain is up to the runtime. Subjectively, I don't think it's reasonable to assume that the probability of the service developer screwing up HTTP API (but not screwing up anything else) is high enough to complicate workflow because of it.

Reply:
Thanks, I agree it is reasonable to not complicate the workflow for highly unlikely scenarious (also given the user tests their code with testkit).